<a href="https://colab.research.google.com/github/alitarraf/Data-Science-Training/blob/master/PandasTraining.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Python Pandas Tutorial: A Complete Introduction for Beginners

![alt text](https://github.com/alitarraf/Data-Science-Training/blob/master/assets/panda%20(2).jpg?raw=true)

The *pandas* package is the most important tool at the disposal of Data Scientists and Analysts working in Python today. The powerful machine learning and glamorous visualization tools may get all the attention, but pandas is the backbone of most data projects. 

>\[*pandas*\] is derived from the term "**pan**el **da**ta", an econometrics term for data sets that include observations over multiple time periods for the same individuals. — [Wikipedia](https://en.wikipedia.org/wiki/Pandas_%28software%29)

If you're thinking about data science as a career, then it is imperative that one of the first things you do is learn pandas. In this post, we will go over the essential bits of information about pandas, including how to install it, its uses, and how it works with other common Python data analysis packages such as **matplotlib** and **sci-kit learn**.

## What's Pandas for?

Pandas has so many uses that it might make sense to list the things it can't do instead of what it can do. 

This tool is essentially your data’s home. Through pandas, you get acquainted with your data by cleaning, transforming, and analyzing it. 

For example, say you want to explore a dataset stored in a CSV on your computer. Pandas will extract the data from that CSV into a DataFrame — a table, basically — then let you do things like:

- Calculate statistics and answer questions about the data, like


    - What's the average, median, max, or min of each column? 
    - Does column A correlate with column B?
    - What does the distribution of data in column C look like?


- Clean the data by doing things like removing missing values and filtering rows or columns by some criteria


- Visualize the data with help from Matplotlib. Plot bars, lines, histograms, bubbles, and more. 


- Store the cleaned, transformed data back into a CSV, other file or database


Before you jump into the modeling or the complex visualizations you need to have a good understanding of the nature of your dataset and pandas is the best avenue through which to do that.



## How does pandas fit into the data science toolkit?

Not only is the pandas library a central component of the data science toolkit but it is used in conjunction with other libraries in that collection. 

Pandas is built on top of the **NumPy** package, meaning a lot of the structure of NumPy is used or replicated in Pandas. Data in pandas is often used to feed statistical analysis in **SciPy**, plotting functions from **Matplotlib**, and machine learning algorithms in **Scikit-learn**.

Jupyter Notebooks offer a good environment for using pandas to do data exploration and modeling, but pandas can also be used in text editors just as easily.

Jupyter Notebooks give us the ability to execute code in a particular cell as opposed to running the entire file. This saves a lot of time when working with large datasets and complex transformations. Notebooks also provide an easy way to visualize pandas’ DataFrames and plots. As a matter of fact, this article was created entirely in a Jupyter Notebook.



## When should you start using pandas?

If you do not have any experience coding in Python, then you should stay away from learning pandas until you do. You don’t have to be at the level of the software engineer, but you should be adept at the basics, such as lists, tuples, dictionaries, functions, and iterations. Also, I’d also recommend familiarizing yourself with **NumPy** due to the similarities mentioned above.

If you're looking for a good place to learn Python, [Python for Everybody](https://www.learndatasci.com/out/coursera-programming-everybody-getting-started-python/) on Coursera is great (and Free).

Moreover, for those of you looking to do a [data science bootcamp](https://www.learndatasci.com/articles/thinkful-data-science-online-bootcamp-review/) or some other accelerated data science education program, it's highly recommended you start learning pandas on your own before you start the program. 

Even though accelerated programs teach you pandas, better skills beforehand means you'll be able to maximize time for learning and mastering the more complicated material.

## Downloading Files

For simplicity and gaining time, let's run the code to downlaod files:



In [0]:
!git clone https://github.com/alitarraf/Data-Science-Training.git  # download data from Github
# change directory into the downloaded folder
%cd "Data-Science-Training/data"
!ls -lah  # list files in current directory

Cloning into 'Data-Science-Training'...
remote: Enumerating objects: 66, done.
remote: Counting objects: 100% (66/66), done.
remote: Compressing objects: 100% (62/62), done.
remote: Total 66 (delta 11), reused 47 (delta 1), pack-reused 0
Unpacking objects: 100% (66/66), done.
/content/Data-Science-Training/data
total 12M
drwxr-xr-x 3 root root  4.0K Jun 20 17:01  .
drwxr-xr-x 5 root root  4.0K Jun 20 17:01  ..
-rw-r--r-- 1 root root   18K Jun 20 17:01  Admission_Predict_Ver1.2.csv
-rw-r--r-- 1 root root  8.8M Jun 20 17:01  Barcelona_Listing_2019.csv
-rw-r--r-- 1 root root   31K Jun 20 17:01 'diet - medium.xls'
-rw-r--r-- 1 root root  302K Jun 20 17:01  IMDB-Movie-Data.csv
-rw-r--r-- 1 root root 1004K Jun 20 17:01  Lebanese_Governorates.json
-rw-r--r-- 1 root root   640 Jun 20 17:01  LebanonPopulation.csv
drwxr-xr-x 2 root root  4.0K Jun 20 17:01 'Lebanon Shapefile'
-rw-r--r-- 1 root root    55 Jun 20 17:01  purchases.csv
-rw-r--r-- 1 root root  9.7K Jun 20 17:01  purchases.xlsx
-rw-r--

## Pandas First Steps



### Install and import
Pandas is an easy package to install. Open up your terminal program (for Mac users) or command line (for PC users) and install it using either of the following commands:

`conda install pandas`

OR 

`pip install pandas`

Alternatively, if you're currently viewing this article in a Jupyter notebook you can run this cell:

In [0]:
!pip install pandas

The `!` at the beginning runs cells as if they were in a terminal.

To import pandas we usually import it with a shorter name since it's used so much:

In [0]:
import pandas as pd

Now to the basic components of pandas.

## Core components of pandas: Series and DataFrames


A `Series` is essentially a column.

A `DataFrame` is a multi-dimensional table made up of a collection of Series. 


DataFrames and Series are quite similar in that many operations that you can do with one you can do with the other, such as filling in null values and calculating the mean.

You'll see how these components work when we start working with data below. 

### Creating DataFrames from scratch
Creating DataFrames right in Python is good to know and quite useful when testing new methods and functions you find in the pandas docs.

There are *many* ways to create a DataFrame from scratch, but a great option is to just use a simple `dict`. 

Let's say we have a fruit stand that sells apples and oranges. We want to have a column for each fruit and a row for each customer purchase. To organize this as a dictionary for pandas we could do something like:

In [0]:
data = {
    'apples': [3, 2, 0, 1], 
    'oranges': [0, 3, 7, 2]
}

And then pass it to the pandas DataFrame constructor:

In [0]:
purchases = pd.DataFrame(data)

purchases

apples  oranges
0       3        0
1       2        3
2       0        7
3       1        2

**How did that work?**

Each *(key, value)* item in `data` corresponds to a *column* in the resulting DataFrame.

The **Index** of this DataFrame was given to us on creation as the numbers 0-3, but we could also create our own when we initialize the DataFrame. 

Let's have customer names as our index: 

In [0]:
purchases = pd.DataFrame(data, index=['June', 'Robert', 'Lily', 'David'])

purchases

apples  oranges
June         3        0
Robert       2        3
Lily         0        7
David        1        2

So now we could **loc**ate a customer's order by using their name as we will see later.

## How to read in data

It’s quite simple to load data from various file formats into a DataFrame. In the following examples we'll keep using our apples and oranges data, but this time it's coming from various files.

### Reading data from Excel
With Excel files all you need is a single line to load in the data:

In [0]:
df = pd.read_excel('purchases.xlsx')
df

Unnamed: 0  apples  oranges
0       June       3        0
1     Robert       2        3
2       Lily       0        7
3      David       1        2

XLSXs don't have indexes like our DataFrames, so all we need to do is just designate the index_col when reading:

In [0]:
df = pd.read_excel('purchases.xlsx', index_col=0)

df

apples  oranges
June         3        0
Robert       2        3
Lily         0        7
David        1        2

### Reading data from CSVs

With CSV files all you need is a single line to load in the data:

Read documentation of "read csv" [here](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html).

In [0]:
# Read file 'purchases.csv'

### START CODE HERE ### (aprox. 1 line of code)
df = pd.read_csv('purchases.csv')
### END CODE HERE ###

In [0]:
df

Unnamed: 0  apples  oranges
0       June       3        0
1     Robert       2        3
2       Lily       0        7
3      David       1        2

Expected Output:

![alt text](https://github.com/alitarraf/Data-Science-Training/blob/master/assets/Ex1.png?raw=true)

CSVs don't have indexes like our DataFrames, so all we need to do is just designate the `index_col` when reading:

In [0]:
# Set column 0 as index

### START CODE HERE ### (aprox. 1 line of code)
df = pd.read_csv('purchases.csv', index_col=0)
### END CODE HERE ###

In [0]:
df

apples  oranges
June         3        0
Robert       2        3
Lily         0        7
David        1        2

Expected Output:

![alt text](https://github.com/alitarraf/Data-Science-Training/blob/master/assets/Ex2.png?raw=true)

Here we're setting the index to be column zero.

You'll find that most CSVs won't ever have an index column and so usually you don't have to worry about this step.


### Reading data from JSON

If you have a JSON file — which is essentially a stored Python `dict` — pandas can read this just as easily:

In [0]:
df = pd.read_json('purchases.json')

df

apples  oranges
David        1        2
June         3        0
Lily         0        7
Robert       2        3

Notice this time our index came with us correctly since using JSON allowed indexes to work through nesting. Feel free to open `data_file.json` in a notepad so you can see how it works.

Pandas will try to figure out how to create a DataFrame by analyzing structure of your JSON, and sometimes it doesn't get it right. Often you'll need to set the `orient` keyword argument depending on the structure, so check out [read_json docs](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_json.html) about that argument to see which orientation you're using.


### Reading data from a SQL database

If you’re working with data from a SQL database you need to first establish a connection using an appropriate Python library, then pass a query to pandas. Here we'll use SQLite to demonstrate. 

First, we need `pysqlite3` installed, so run this command in your terminal:

`pip install pysqlite3`

Or run this cell if you're in a notebook:

In [0]:
!pip install pysqlite3

`sqlite3` is used to create a connection to a database which we can then use to generate a DataFrame through a `SELECT` query.

So first we'll make a connection to a SQLite database file:

In [0]:
import sqlite3

con = sqlite3.connect("database.db")

**Note**: If you have data in PostgreSQL, MySQL, or some other SQL server, you'll need to obtain the right Python library to make a connection. For example, `psycopg2` ([link](http://initd.org/psycopg/download/)) is a commonly used library for making connections to PostgreSQL. Furthermore, you would make a connection to a database URI instead of a file like we did here with SQLite. For a great course on SQL check out [The Complete SQL Bootcamp](https://learndatasci.com/out/udemy-the-complete-sql-bootcamp/) on Udemy


In this SQLite database we have a table called *purchases*, and our index is in a column called "index". 

By passing a SELECT query and our `con`, we can read from the *purchases* table:

In [0]:
df = pd.read_sql_query("SELECT * FROM purchases", con)

df

index  apples  oranges
0    June       3        0
1  Robert       2        3
2    Lily       0        7
3   David       1        2

Just like with CSVs, we could pass `index_col='index'`, but we can also set an index after-the-fact:

In [0]:
df = df.set_index('index')

df

apples  oranges
index                  
June         3        0
Robert       2        3
Lily         0        7
David        1        2

In fact, we could use `set_index()` on *any* DataFrame using *any* column at *any* time. Indexing Series and DataFrames is a very common task, and the different ways of doing it is worth remembering.

### Converting back to a XLSX, CSV, JSON, or SQL

So after extensive work on cleaning your data, you’re now ready to save it as a file of your choice. Similar to the ways we read in data, pandas provides intuitive commands to save it:

In [0]:
df.to_excel('new_purchases.xlsx')

df.to_csv('new_purchases.csv')

# df.to_json('new_purchases.json')

# df.to_sql('new_purchases', con)

When we save XLSX, JSON and CSV files, all we have to input into those functions is our desired filename with the appropriate file extension. With SQL, we’re not creating a new file but instead inserting a new table into the database using our `con` variable from before.

Let's move on to importing some real-world data and detailing a few of the operations you'll be using a lot.

## Most important DataFrame operations

DataFrames possess hundreds of methods and other operations that are crucial to any analysis. As a beginner, you should know the operations that perform simple transformations of your data and those that provide fundamental statistical analysis.

Let's load in the IMDB movies dataset to begin:

In [0]:
# Read IMDB movies CSV file, save it in "movies_df", and set Title as column index
#To see output, refer to next section.

### START CODE HERE ### (aprox. 1 line of code)
movies_df = pd.read_csv("IMDB-Movie-Data.csv", index_col="Title")
### END CODE HERE ###

We're loading this dataset from a CSV and designating the movie titles to be our index.


### Viewing your data

The first thing to do when opening a new dataset is print out a few rows to keep as a visual reference. We accomplish this with `.head()`:

In [0]:
movies_df.head()

Rank  ... Metascore
Title                          ...          
Guardians of the Galaxy     1  ...      76.0
Prometheus                  2  ...      65.0
Split                       3  ...      62.0
Sing                        4  ...      59.0
Suicide Squad               5  ...      40.0

[5 rows x 11 columns]

Expected Output:

![alt text](https://github.com/alitarraf/Data-Science-Training/blob/master/assets/exx.jpg?raw=true)


`.head()` outputs the **first** five rows of your DataFrame by default, but we could also pass a number as well: `movies_df.head(10)` would output the top ten rows, for example. 

To see the **last** five rows use `.tail()`. 

`tail()` also accepts a number, and in this case we printing the bottom two rows.:

In [0]:
# Show the last 2 rows of movies_df

### START CODE HERE ### (approx. 1 line of code)
movies_df.tail(2)
### END CODE HERE ###

Rank                  Genre  ... Revenue (Millions) Metascore
Title                                      ...                             
Search Party   999       Adventure,Comedy  ...                NaN      22.0
Nine Lives    1000  Comedy,Family,Fantasy  ...              19.64      11.0

[2 rows x 11 columns]

Expected Output:

![alt text](https://github.com/alitarraf/Data-Science-Training/blob/master/assets/Ex3.png?raw=true)

Typically when we load in a dataset, we like to view the first five or so rows to see what's under the hood. Here we can see the names of each column, the index, and examples of values in each row.

You'll notice that the index in our DataFrame is the *Title* column, which you can tell by how the word *Title* is slightly lower than the rest of the columns.

### Getting info about your data

`.info()` should be one of the very first commands you run after loading your data:

In [0]:
movies_df.info()

type(movies_df['Actors'][3])

<class 'pandas.core.frame.DataFrame'>
Index: 1000 entries, Guardians of the Galaxy to Nine Lives
Data columns (total 11 columns):
Rank                  1000 non-null int64
Genre                 1000 non-null object
Description           1000 non-null object
Director              1000 non-null object
Actors                1000 non-null object
Year                  1000 non-null int64
Runtime (Minutes)     1000 non-null int64
Rating                1000 non-null float64
Votes                 1000 non-null int64
Revenue (Millions)    872 non-null float64
Metascore             936 non-null float64
dtypes: float64(3), int64(4), object(4)
memory usage: 93.8+ KB


str

`.info()` provides the essential details about your dataset, such as the number of rows and columns, the number of non-null values, what type of data is in each column, and how much memory your DataFrame is using. 

Notice in our movies dataset we have some obvious missing values in the `Revenue` and `Metascore` columns. We'll look at how to handle those in a bit.

Seeing the datatype quickly is actually quite useful. Imagine you just imported some JSON and the integers were recorded as strings. You go to do some arithmetic and find an "unsupported operand" Exception because you can't do math with strings. Calling `.info()` will quickly point out that your column you thought was all integers are actually string objects.

Another fast and useful attribute is `.shape`, which outputs just a tuple of (rows, columns):

In [0]:
movies_df.shape

(1000, 11)

Note that `.shape` has no parentheses and is a simple tuple of format (rows, columns). So we have **1000 rows** and **11 columns** in our movies DataFrame.

You'll be going to `.shape` a lot when cleaning and transforming data. For example, you might filter some rows based on some criteria and then want to know quickly how many rows were removed.

### Handling duplicates

This dataset does not have duplicate rows, but it is always important to verify you aren't aggregating duplicate rows. 

To demonstrate, let's simply just double up our movies DataFrame by appending it to itself:

In [0]:
temp_df = movies_df.append(movies_df)

temp_df.shape

(2000, 11)

Using `append()` will return a copy without affecting the original DataFrame. We are capturing this copy in `temp` so we aren't working with the real data.

Notice call `.shape` quickly proves our DataFrame rows have doubled.

Now we can try dropping duplicates:

In [0]:
temp_df = temp_df.drop_duplicates()

temp_df.shape

(1000, 11)

Just like `append()`, the `drop_duplicates()` method will also return a copy of your DataFrame, but this time with duplicates removed. Calling `.shape` confirms we're back to the 1000 rows of our original dataset.

It's a little verbose to keep assigning DataFrames to the same variable like in this example. For this reason, pandas has the `inplace` keyword argument on many of its methods. Using `inplace=True` will modify the DataFrame object in place:

In [0]:
temp_df.drop_duplicates(inplace=True)

Now our `temp_df` *will* have the transformed data automatically. 

Another important argument for `drop_duplicates()` is `keep`, which has three possible options:

* `first`: (default) Drop duplicates except for the first occurrence.
* `last`: Drop duplicates except for the last occurrence.
* `False`: Drop all duplicates.

Since we didn't define the `keep` arugment in the previous example it was defaulted to `first`. This means that if two rows are the same pandas will drop the second row and keep the first row. Using `last` has the opposite effect: the first row is dropped.

`False`, on the other hand, will drop all duplicates. If two rows are the same then both will be dropped. Watch what happens to `temp_df`:

In [0]:
# Duplicate 'movies_df' and save the new dataframe as 'temp_df'
# Drop ALL duplicates of 'temp_df' and overwrite it with the new dataframe
# Output shape of 'temp_df'

### START CODE HERE ### (approx. 3 lines of code)
temp_df = movies_df.append(movies_df)  # make a new copy
temp_df.drop_duplicates(inplace=True, keep=False)
temp_df.shape
### END CODE HERE ###

(0, 11)

Expected Output:

(0, 11)

Since all rows were duplicates, `keep=False` dropped them all resulting in zero rows being left over. If you're wondering why you would want to do this, one reason is that it allows you to locate all duplicates in your dataset. When conditional selections are shown below you'll see how to do that.

### Column cleanup

Many times datasets will have verbose column names with symbols, upper and lowercase words, spaces, and typos. To make selecting data by column name easier we can spend a little time cleaning up their names.

Here's how to print the column names of our dataset:

In [0]:
movies_df.columns

Index(['Rank', 'Genre', 'Description', 'Director', 'Actors', 'Year',
       'Runtime (Minutes)', 'Rating', 'Votes', 'Revenue (Millions)',
       'Metascore'],
      dtype='object')

Not only does `.columns` come in handy if you want to rename columns by allowing for simple copy and paste, it's also useful if you need to understand why you are receiving a `Key Error` when selecting data by column.

We can use the `.rename()` method to rename certain or all columns via a `dict`. We don't want parentheses, so let's rename those:

In [0]:
movies_df.rename(columns={
        'Runtime (Minutes)': 'Runtime', 
        'Revenue (Millions)': 'Revenue_millions'
    }, inplace=True)


movies_df.columns

Index(['Rank', 'Genre', 'Description', 'Director', 'Actors', 'Year', 'Runtime',
       'Rating', 'Votes', 'Revenue_millions', 'Metascore'],
      dtype='object')

Excellent. But what if we want to lowercase all names? Instead of using `.rename()` we could also set a list of names to the columns like so:

In [0]:
movies_df.columns = ['rank', 'genre', 'description', 'director', 'actors', 'year', 'runtime', 
                     'rating', 'votes', 'revenue_millions', 'metascore']


movies_df.columns

Index(['rank', 'genre', 'description', 'director', 'actors', 'year', 'runtime',
       'rating', 'votes', 'revenue_millions', 'metascore'],
      dtype='object')

But that's too much work. Instead of just renaming each column manually we can do a list comprehension:

In [0]:
movies_df.columns = [col.lower() for col in movies_df]
#can change "col" ==> [x.lower() for x in movies_df]

movies_df.columns

Index(['rank', 'genre', 'description', 'director', 'actors', 'year', 'runtime',
       'rating', 'votes', 'revenue_millions', 'metascore'],
      dtype='object')

`list` (and `dict`) comprehensions come in handy a lot when working with pandas and data in general.

It's a good idea to lowercase, remove special characters, and replace spaces with underscores if you'll be working with a dataset for some time.

### How to work with missing values

When exploring data, you’ll most likely encounter missing or null values, which are essentially placeholders for non-existent values. Most commonly you'll see Python's `None` or NumPy's `np.nan`, each of which are handled differently in some situations.

There are two options in dealing with nulls: 

1. Get rid of rows or columns with nulls
2. Replace nulls with non-null values, a technique known as **imputation**

Let's calculate to total number of nulls in each column of our dataset. The first step is to check which cells in our DataFrame are null:

In [0]:
movies_df.isnull()

rank  ...  metascore
Title                                           ...           
Guardians of the Galaxy                  False  ...      False
Prometheus                               False  ...      False
Split                                    False  ...      False
Sing                                     False  ...      False
Suicide Squad                            False  ...      False
The Great Wall                           False  ...      False
La La Land                               False  ...      False
Mindhorn                                 False  ...      False
The Lost City of Z                       False  ...      False
Passengers                               False  ...      False
Fantastic Beasts and Where to Find Them  False  ...      False
Hidden Figures                           False  ...      False
Rogue One                                False  ...      False
Moana                                    False  ...      False
Colossal                                 False  ...      False
The Secret Life of Pets                  False  ...      False
Hacksaw Ridge                            False  ...      False
Jason Bourne                             False  ...      False
Lion                                     False  ...      False
Arrival                                  False  ...      False
Gold                                     False  ...      False
Manchester by the Sea                    False  ...      False
Hounds of Love                           False  ...      False
Trolls                                   False  ...      False
Independence Day: Resurgence             False  ...      False
Paris pieds nus                          False  ...       True
Bahubali: The Beginning                  False  ...       True
Dead Awake                               False  ...       True
Bad Moms                                 False  ...      False
Assassin's Creed                         False  ...      False
...                                        ...  ...        ...
Texas Chainsaw 3D                        False  ...      False
Disturbia                                False  ...       True
Rock of Ages                             False  ...      False
Scream 4                                 False  ...      False
Queen of Katwe                           False  ...      False
My Big Fat Greek Wedding 2               False  ...      False
Dark Places                              False  ...      False
Amateur Night                            False  ...      False
It's Only the End of the World           False  ...      False
The Skin I Live In                       False  ...      False
Miracles from Heaven                     False  ...      False
Annie                                    False  ...      False
Across the Universe                      False  ...      False
Let's Be Cops                            False  ...      False
Max                                      False  ...      False
Your Highness                            False  ...      False
Final Destination 5                      False  ...      False
Endless Love                             False  ...      False
Martyrs                                  False  ...      False
Selma                                    False  ...       True
Underworld: Rise of the Lycans           False  ...      False
Taare Zameen Par                         False  ...      False
Take Me Home Tonight                     False  ...       True
Resident Evil: Afterlife                 False  ...      False
Project X                                False  ...      False
Secret in Their Eyes                     False  ...      False
Hostel: Part II                          False  ...      False
Step Up 2: The Streets                   False  ...      False
Search Party                             False  ...      False
Nine Lives                               False  ...      False

[1000 rows x 11 columns]

Notice `isnull()` returns a DataFrame where each cell is either True or False depending on that cell's null status.

To count the number of nulls in each column we use an aggregate function for summing: 

In [0]:
movies_df.isnull().sum()

rank                  0
genre                 0
description           0
director              0
actors                0
year                  0
runtime               0
rating                0
votes                 0
revenue_millions    128
metascore            64
dtype: int64

`.isnull()` just by iteself isn't very useful, and is usually used in conjunction with other methods, like `sum()`.

We can see now that our data has **128** missing values for `revenue_millions` and **64** missing values for `metascore`.

#### Removing null values

Data Scientists and Analysts regularly face the dilemma of dropping or imputing null values, and is a decision that requires intimate knowledge of your data and its context. Overall, removing null data is only suggested if you have a small amount of missing data.

Remove nulls is pretty simple:

In [0]:
movies_df.dropna()

rank  ... metascore
Title                                          ...          
Guardians of the Galaxy                     1  ...      76.0
Prometheus                                  2  ...      65.0
Split                                       3  ...      62.0
Sing                                        4  ...      59.0
Suicide Squad                               5  ...      40.0
The Great Wall                              6  ...      42.0
La La Land                                  7  ...      93.0
The Lost City of Z                          9  ...      78.0
Passengers                                 10  ...      41.0
Fantastic Beasts and Where to Find Them    11  ...      66.0
Hidden Figures                             12  ...      74.0
Rogue One                                  13  ...      65.0
Moana                                      14  ...      81.0
Colossal                                   15  ...      70.0
The Secret Life of Pets                    16  ...      61.0
Hacksaw Ridge                              17  ...      71.0
Jason Bourne                               18  ...      58.0
Lion                                       19  ...      69.0
Arrival                                    20  ...      81.0
Gold                                       21  ...      49.0
Manchester by the Sea                      22  ...      96.0
Trolls                                     24  ...      56.0
Independence Day: Resurgence               25  ...      32.0
Bad Moms                                   29  ...      60.0
Assassin's Creed                           30  ...      36.0
Why Him?                                   31  ...      39.0
Nocturnal Animals                          32  ...      67.0
X-Men: Apocalypse                          33  ...      52.0
Deadpool                                   34  ...      65.0
Resident Evil: The Final Chapter           35  ...      49.0
...                                       ...  ...       ...
That Awkward Moment                       956  ...      36.0
Legion                                    957  ...      32.0
End of Watch                              958  ...      68.0
3 Days to Kill                            959  ...      40.0
Lucky Number Slevin                       960  ...      53.0
Trance                                    961  ...      61.0
Into the Forest                           962  ...      59.0
The Other Boleyn Girl                     963  ...      50.0
I Spit on Your Grave                      964  ...      27.0
Texas Chainsaw 3D                         971  ...      62.0
Rock of Ages                              973  ...      47.0
Scream 4                                  974  ...      52.0
Queen of Katwe                            975  ...      73.0
My Big Fat Greek Wedding 2                976  ...      37.0
The Skin I Live In                        980  ...      70.0
Miracles from Heaven                      981  ...      44.0
Annie                                     982  ...      33.0
Across the Universe                       983  ...      56.0
Let's Be Cops                             984  ...      30.0
Max                                       985  ...      47.0
Your Highness                             986  ...      31.0
Final Destination 5                       987  ...      50.0
Endless Love                              988  ...      30.0
Underworld: Rise of the Lycans            991  ...      44.0
Taare Zameen Par                          992  ...      42.0
Resident Evil: Afterlife                  994  ...      37.0
Project X                                 995  ...      48.0
Hostel: Part II                           997  ...      46.0
Step Up 2: The Streets                    998  ...      50.0
Nine Lives                               1000  ...      11.0

[838 rows x 11 columns]

This operation will delete any **row** with at least a single null value, but it will return a new DataFrame without altering the original one. You could specify `inplace=True` in this method as well.

So in the case of our dataset, this operation would remove 128 rows where `revenue_millions` is null and 64 rows where `metascore` is null. This obviously seems like a waste since there's perfectly good data in the other columns of those dropped rows. That's why we'll look at imputation next.

Other than just dropping rows, you can also drop columns with null values by setting `axis=1`:

In [0]:
movies_df.dropna(axis=1)

rank  ...   votes
Title                                          ...        
Guardians of the Galaxy                     1  ...  757074
Prometheus                                  2  ...  485820
Split                                       3  ...  157606
Sing                                        4  ...   60545
Suicide Squad                               5  ...  393727
The Great Wall                              6  ...   56036
La La Land                                  7  ...  258682
Mindhorn                                    8  ...    2490
The Lost City of Z                          9  ...    7188
Passengers                                 10  ...  192177
Fantastic Beasts and Where to Find Them    11  ...  232072
Hidden Figures                             12  ...   93103
Rogue One                                  13  ...  323118
Moana                                      14  ...  118151
Colossal                                   15  ...    8612
The Secret Life of Pets                    16  ...  120259
Hacksaw Ridge                              17  ...  211760
Jason Bourne                               18  ...  150823
Lion                                       19  ...  102061
Arrival                                    20  ...  340798
Gold                                       21  ...   19053
Manchester by the Sea                      22  ...  134213
Hounds of Love                             23  ...    1115
Trolls                                     24  ...   38552
Independence Day: Resurgence               25  ...  127553
Paris pieds nus                            26  ...     222
Bahubali: The Beginning                    27  ...   76193
Dead Awake                                 28  ...     523
Bad Moms                                   29  ...   66540
Assassin's Creed                           30  ...  112813
...                                       ...  ...     ...
Texas Chainsaw 3D                         971  ...   37060
Disturbia                                 972  ...  193491
Rock of Ages                              973  ...   64513
Scream 4                                  974  ...  108544
Queen of Katwe                            975  ...    6753
My Big Fat Greek Wedding 2                976  ...   20966
Dark Places                               977  ...   31634
Amateur Night                             978  ...    2229
It's Only the End of the World            979  ...   10658
The Skin I Live In                        980  ...  108772
Miracles from Heaven                      981  ...   12048
Annie                                     982  ...   27312
Across the Universe                       983  ...   95172
Let's Be Cops                             984  ...  112729
Max                                       985  ...   21405
Your Highness                             986  ...   87904
Final Destination 5                       987  ...   88000
Endless Love                              988  ...   33688
Martyrs                                   989  ...   63785
Selma                                     990  ...   67637
Underworld: Rise of the Lycans            991  ...  129708
Taare Zameen Par                          992  ...  102697
Take Me Home Tonight                      993  ...   45419
Resident Evil: Afterlife                  994  ...  140900
Project X                                 995  ...  164088
Secret in Their Eyes                      996  ...   27585
Hostel: Part II                           997  ...   73152
Step Up 2: The Streets                    998  ...   70699
Search Party                              999  ...    4881
Nine Lives                               1000  ...   12435

[1000 rows x 9 columns]

In our dataset, this operation would drop the `revenue_millions` and `metascore` columns.

**Intuition side note**: What's with this `axis=1` parameter?

It's not immediately obvious where `axis` comes from and why you need it to be 1 for it to affect columns. To see why, just look at the `.shape` output:

In [0]:
movies_df.shape

(1000, 11)

As we learned above, this is a tuple that represents the shape of the DataFrame, i.e. 1000 rows and 11 columns. Note that the *rows* are at index zero of this tuple and *columns* are at **index one** of this tuple. This is why `axis=1` affects columns. This comes from NumPy, and is a great example of why learning NumPy is worth your time.

#### Imputation

Imputation is a conventional feature engineering technique used to keep valuable data that have null values. 

There may be instances where dropping every row with a null value removes too big a chunk from your dataset, so instead we can impute that null with another value, usually the **mean** or the **median** of that column. 

Let's look at imputing the missing values in the `revenue_millions` column. First we'll extract that column into its own variable:

In [0]:
revenue = movies_df['revenue_millions']

Using square brackets is the general way we select columns in a DataFrame. 

If you remember back to when we created DataFrames from scratch, the keys of the `dict` ended up as column names. Now when we select columns of a DataFrame, we use brackets just like if we were accessing a Python dictionary. 

`revenue` now contains a Series:

In [0]:
revenue.head()

Title
Guardians of the Galaxy    333.13
Prometheus                 126.46
Split                      138.12
Sing                       270.32
Suicide Squad              325.02
Name: revenue_millions, dtype: float64

Slightly different formatting than a DataFrame, but we still have our `Title` index. 

We'll impute the missing values of revenue using the mean. With the mean, let's fill the nulls using fillna(). By using `inplace=True` we can actually affect the original `movies_df`

In [0]:
# Fill null values of revenue with its mean and use inplace to ovewrite revenue
# Note htat the mean of revenue is "revenue.mean"
# Check if the sum of null values for each column is 0

### START CODE HERE ### (aprox. 2 lines of code)
revenue.fillna(revenue.mean(), inplace=True)
movies_df.isnull().sum()
### END CODE HERE ###

rank                 0
genre                0
description          0
director             0
actors               0
year                 0
runtime              0
rating               0
votes                0
revenue_millions     0
metascore           64
dtype: int64

Expected Output:

![alt text](https://github.com/alitarraf/Data-Science-Training/blob/master/assets/Ex4.png?raw=true)

Imputing an entire column with the same value like this is a basic example. It would be a better idea to try a more granular imputation by Genre or Director. 

For example, you would find the mean of the revenue generated in each genre individually and impute the nulls in each genre with that genre's mean.

Let's now look at more ways to examine and understand the dataset.

### Understanding your variables

Using `describe()` on an entire DataFrame we can get a summary of the distribution of continuous variables:

In [0]:
movies_df.describe()

rank         year  ...  revenue_millions   metascore
count  1000.000000  1000.000000  ...       1000.000000  936.000000
mean    500.500000  2012.783000  ...         82.956376   58.985043
std     288.819436     3.205962  ...         96.412043   17.194757
min       1.000000  2006.000000  ...          0.000000   11.000000
25%     250.750000  2010.000000  ...         17.442500   47.000000
50%     500.500000  2014.000000  ...         60.375000   59.500000
75%     750.250000  2016.000000  ...         99.177500   72.000000
max    1000.000000  2016.000000  ...        936.630000  100.000000

[8 rows x 7 columns]

Understanding which numbers are continuous also comes in handy when thinking about the type of plot to use to represent your data visually. 

`.describe()` can also be used on a categorical variable to get the count of rows, unique count of categories, top category, and freq of top category:

In [0]:
movies_df['genre'].describe()

count                        1000
unique                        207
top       Action,Adventure,Sci-Fi
freq                           50
Name: genre, dtype: object

This tells us that the genre column has 207 unique values, the top value is Action/Adventure/Sci-Fi, which shows up 50 times (freq).

`.value_counts()` can tell us the frequency of all values in a column:

In [0]:
movies_df['genre'].value_counts().head(10)

Action,Adventure,Sci-Fi       50
Drama                         48
Comedy,Drama,Romance          35
Comedy                        32
Drama,Romance                 31
Action,Adventure,Fantasy      27
Comedy,Drama                  27
Animation,Adventure,Comedy    27
Comedy,Romance                26
Crime,Drama,Thriller          24
Name: genre, dtype: int64

#### Relationships between continuous variables

By using the correlation method `.corr()` we can generate the relationship between each continuous variable:

In [0]:
movies_df.corr()

rank      year  ...  revenue_millions  metascore
rank              1.000000 -0.261605  ...         -0.252996  -0.191869
year             -0.261605  1.000000  ...         -0.117562  -0.079305
runtime          -0.221739 -0.164900  ...          0.247834   0.211978
rating           -0.219555 -0.211219  ...          0.189527   0.631897
votes            -0.283876 -0.411904  ...          0.607941   0.325684
revenue_millions -0.252996 -0.117562  ...          1.000000   0.133328
metascore        -0.191869 -0.079305  ...          0.133328   1.000000

[7 rows x 7 columns]

Correlation tables are a numerical representation of the bivariate relationships in the dataset. 

Positive numbers indicate a positive correlation — one goes up the other goes up — and negative numbers represent an inverse correlation — one goes up the other goes down. 1.0 indicates a perfect correlation. 

So looking in the first row, first column we see `rank` has a perfect correlation with itself, which is obvious. On the other hand, the correlation between `votes` and `revenue_millions` is 0.6. A little more interesting.

Examining bivariate relationships comes in handy when you have an outcome or dependent variable in mind and would like to see the features most correlated to the increase or decrease of the outcome. You can visually represent bivariate relationships with scatterplots (seen below in the plotting section). 

For a deeper look into data summarizations check out [Essential Statistics for Data Science](https://www.learndatasci.com/tutorials/data-science-statistics-using-python/).

Let's now look more at manipulating DataFrames.

### DataFrame slicing, selecting, extracting

Up until now we've focused on some basic summaries of our data. We've learned about simple column extraction using single brackets, and we imputed null values in a column using `fillna()`. Below are the other methods of slicing, selecting, and extracting you'll need to use constantly.

It's important to note that, although many methods are the same, DataFrames and Series have different attributes, so you'll need be sure to know which type you are working with or else you will receive attribute errors. 

Let's look at working with columns first.

#### By column

You already saw how to extract a column using square brackets like this:

In [0]:
genre_col = movies_df['genre']

type(genre_col)

pandas.core.series.Series

This will return a *Series*. To extract a column as a *DataFrame*, you need to pass a list of column names. In our case that's just a single column:

In [0]:
genre_col = movies_df[['genre']]

type(genre_col)

pandas.core.frame.DataFrame

Since it's just a list, adding another column name is easy:

In [0]:
subset = movies_df[['genre', 'rating']]

subset.head()

genre  rating
Title                                                    
Guardians of the Galaxy   Action,Adventure,Sci-Fi     8.1
Prometheus               Adventure,Mystery,Sci-Fi     7.0
Split                             Horror,Thriller     7.3
Sing                      Animation,Comedy,Family     7.2
Suicide Squad            Action,Adventure,Fantasy     6.2

Now we'll look at getting data by rows.

#### By rows

For rows, we have two options: 

- `.loc` - **loc**ates by name
- `.iloc`- **loc**ates by numerical **i**ndex

Remember that we are still indexed by movie Title, so to use `.loc` we give it the Title of a movie:

In [0]:
prom = movies_df.loc["Prometheus"]

prom

rank                                                                2
genre                                        Adventure,Mystery,Sci-Fi
description         Following clues to the origin of mankind, a te...
director                                                 Ridley Scott
actors              Noomi Rapace, Logan Marshall-Green, Michael Fa...
year                                                             2012
runtime                                                           124
rating                                                              7
votes                                                          485820
revenue_millions                                               126.46
metascore                                                          65
Name: Prometheus, dtype: object

On the other hand, with `iloc` we give it the numerical index of Prometheus:

In [0]:
prom = movies_df.iloc[1]

`loc` and `iloc` can be thought of as similar to Python `list` slicing. To show this even further, let's select multiple rows.

How would you do it with a list? In Python, just slice with brackets like `example_list[1:4]`. It's works the same way in pandas:

In [0]:
movie_subset = movies_df.loc['Prometheus':'Sing']

movie_subset = movies_df.iloc[1:4]

movie_subset

rank                     genre  ... revenue_millions metascore
Title                                       ...                           
Prometheus     2  Adventure,Mystery,Sci-Fi  ...           126.46      65.0
Split          3           Horror,Thriller  ...           138.12      62.0
Sing           4   Animation,Comedy,Family  ...           270.32      59.0

[3 rows x 11 columns]

One important distinction between using `.loc` and `.iloc` to select multiple rows is that `.loc` includes the movie *Sing* in the result, but when using `.iloc` we're getting rows 1:4 but the movie at index 4 (*Suicide Squad*) is not included. 

Slicing with `.iloc` follows the same rules as slicing with lists, the object at the index at the end is not included.

#### Conditional selections
We’ve gone over how to select columns and rows, but what if we want to make a conditional selection? 

For example, what if we want to filter our movies DataFrame to show only films directed by Ridley Scott or films with a rating greater than or equal to 8.0?

To do that, we take a column from the DataFrame and apply a Boolean condition to it. 

In [0]:
# We want to know if the director is Ridley Scott or not.
#Show the result of the first 5 movie titles, name it "condition"

### START CODE HERE ### (aprox. 1 lines of code)
(movies_df['director'] == "Ridley Scott").head()
### END CODE HERE ###


Title
Guardians of the Galaxy    False
Prometheus                  True
Split                      False
Sing                       False
Suicide Squad              False
Name: director, dtype: bool

Expected Output:

![alt text](https://github.com/alitarraf/Data-Science-Training/blob/master/assets/Ex5.png?raw=true)

Similar to `isnull()`, this returns a Series of True and False values: True for films directed by Ridley Scott and False for ones not directed by him. 


In [0]:
#We want to filter out all movies not directed by Ridley Scott, in other words,
#we don’t want the False films.
#Show the result of the first 5 movie titles

### START CODE HERE ### (aprox. 1 lines of code)
movies_df[movies_df['director'] == "Ridley Scott"].head()
### END CODE HERE ###


rank  ... metascore
Title                         ...          
Prometheus                 2  ...      65.0
The Martian              103  ...      80.0
Robin Hood               388  ...      53.0
American Gangster        471  ...      76.0
Exodus: Gods and Kings   517  ...      52.0

[5 rows x 11 columns]

Expected Output:

![alt text](https://github.com/alitarraf/Data-Science-Training/blob/master/assets/Ex6.png?raw=true)

You can get used to looking at these conditionals by reading it like: 

> Select movies_df where movies_df director equals Ridley Scott

Let's look at conditional selections using numerical values by filtering the DataFrame by ratings:

In [0]:
movies_df[movies_df['rating'] >= 8.6].head(3)

rank                    genre  ... revenue_millions metascore
Title                                           ...                           
Interstellar       37   Adventure,Drama,Sci-Fi  ...           187.99      74.0
The Dark Knight    55       Action,Crime,Drama  ...           533.32      82.0
Inception          81  Action,Adventure,Sci-Fi  ...           292.57      74.0

[3 rows x 11 columns]

We can make some richer conditionals by using logical operators `|` for "or" and `&` for "and".

Let's filter the the DataFrame to show only movies by Christopher Nolan OR Ridley Scott:

In [0]:
movies_df[(movies_df['director'] == 'Christopher Nolan') | (movies_df['director'] == 'Ridley Scott')].head()

rank                     genre  ... revenue_millions metascore
Title                                            ...                           
Prometheus          2  Adventure,Mystery,Sci-Fi  ...           126.46      65.0
Interstellar       37    Adventure,Drama,Sci-Fi  ...           187.99      74.0
The Dark Knight    55        Action,Crime,Drama  ...           533.32      82.0
The Prestige       65      Drama,Mystery,Sci-Fi  ...            53.08      66.0
Inception          81   Action,Adventure,Sci-Fi  ...           292.57      74.0

[5 rows x 11 columns]

We need to make sure to group evaluations with parentheses so Python knows how to evaluate the conditional.

Using the `isin()` method we could make this more concise though:

In [0]:
movies_df[movies_df['director'].isin(['Christopher Nolan', 'Ridley Scott'])].head()

rank                     genre  ... revenue_millions metascore
Title                                            ...                           
Prometheus          2  Adventure,Mystery,Sci-Fi  ...           126.46      65.0
Interstellar       37    Adventure,Drama,Sci-Fi  ...           187.99      74.0
The Dark Knight    55        Action,Crime,Drama  ...           533.32      82.0
The Prestige       65      Drama,Mystery,Sci-Fi  ...            53.08      66.0
Inception          81   Action,Adventure,Sci-Fi  ...           292.57      74.0

[5 rows x 11 columns]

Let's say we want all movies that were released between 2005 and 2010, have a rating above 8.0, but made below the 25th percentile in revenue.

Here's how we could do all of that:

## Applying functions

It is possible to iterate over a DataFrame or Series as you would with a list, but doing so — especially on large datasets — is very slow.

An efficient alternative is to `apply()` a function to the dataset. For example, we could use a function to convert movies with an 8.0 or greater to a string value of "good" and the rest to "bad" and use this transformed values to create a new column.

First we would create a function that, when given a rating, determines if it's good or bad:

In [0]:
def rating_function(x):
    if x >= 8.0:
        return "good"
    else:
        return "bad"

Now we want to send the entire rating column through this function, which is what `apply()` does:

In [0]:
movies_df["rating_category"] = movies_df["rating"].apply(rating_function)

movies_df.head(2)

rank  ... rating_category
Title                          ...                
Guardians of the Galaxy     1  ...            good
Prometheus                  2  ...             bad

[2 rows x 12 columns]

The `.apply()` method passes every value in the `rating` column through the `rating_function` and then returns a new Series. This Series is then assigned to a new column called `rating_category`.



In [0]:
# Let's see who will get it all right!
# Using what you just learned, create a function named "acceptance_function" that takes rating category as an input
# This function should return "Accepted" if the rating category is good and "Rejected" otherwise
# Save the results in a new column with title "decision"
# Show ONLY rows of rank 991 to 993 without using column "rank" ;)

### START CODE HERE ### (aprox. 7 lines of code)
def acceptance_function(x):
  if x=="good":
    return "Accepted"
  else:
    return "Rejected"
  
movies_df["decision"] = movies_df["rating_category"].apply(acceptance_function)
movies_df.iloc[-10:-7]
### END CODE HERE ###

rank  ...  decision
Title                                 ...          
Underworld: Rise of the Lycans   991  ...  Rejected
Taare Zameen Par                 992  ...  Accepted
Take Me Home Tonight             993  ...  Rejected

[3 rows x 13 columns]

Expected Output:

![alt text](https://github.com/alitarraf/Data-Science-Training/blob/master/assets/Ex7.PNG?raw=true)

You can also use anonymous functions as well. This lambda function achieves the same result as `rating_function`:

In [0]:
movies_df["rating_category"] = movies_df["rating"].apply(lambda x: 'good' if x >= 8.0 else 'bad')

movies_df.head(2)

rank  ... rating_category
Title                          ...                
Guardians of the Galaxy     1  ...            good
Prometheus                  2  ...             bad

[2 rows x 12 columns]

Overall, using `apply()` will be much faster than iterating manually over rows because pandas is utilizing vectorization.

> Vectorization: a style of computer programming where operations are applied to whole arrays instead of individual elements —[Wikipedia](https://en.wikipedia.org/wiki/Vectorization)

A good example of high usage of `apply()` is during natural language processing (NLP) work. You'll need to apply all sorts of text cleaning functions to strings to prepare for machine learning.

## Wrapping up

Exploring, cleaning, transforming, and visualization data with pandas in Python is an essential skill in data science. Just cleaning wrangling data is 80% of your job as a Data Scientist. After a few projects and some practice, you should be very comfortable with most of the basics.

To keep improving, view the [extensive tutorials](https://pandas.pydata.org/pandas-docs/stable/tutorials.html) offered by the official pandas docs, follow along with a few [Kaggle kernels](https://www.kaggle.com/kernels), and keep working on your own projects!

## References

[Python Pandas Tutorial: A Complete Introduction for Beginnerst](https://www.learndatasci.com/tutorials/python-pandas-tutorial-complete-introduction-for-beginners/)